### PR TITLE
fix list patterns

### PR DESCRIPTION
**Pull request by skogsbaer**
Don't merge, blocked by #1

### CODE REVIEW

**Comment by skogsbaer** [comment]
@albsch This PR is ready for review.

One question: in `subtil.erl`, I find the following

~~~erlang
-spec is_subty(symtab:t(), ast:ty(), ast:ty()) -> boolean().
is_subty(_Symtab, T1, T2) ->
% TODO use symtab
~~~

What's the plan here? We need to symtab for expanding type definitions, right? Does it mean that the code currently does not work with user defined types?

File: src/constr_gen.erl [hunk 1]
```diff
@@ -4,7 +4,10 @@
 
 -export([
          gen_constrs_fun_group/1, gen_constrs_annotated_fun/2,
-         sanity_check/2
+         sanity_check/2,
+         % for tests
+         pat_guard_upper_lower/3,
+         ty_of_pat/3
         ]).
```
Comment (albsch) [inline src/constr_gen.erl:11]: If this is meant just for testing we could wrap it into an ifdef.

```suggestion
        ]).

-ifdef(TEST).
-export([
         pat_guard_upper_lower/3,
         ty_of_pat/3
        ]).
-endif.
```

File: src/constr_gen.erl [hunk 2]
```diff
@@ -324,32 +331,86 @@ bound_vars_pat(P) ->
...
-        {match, _L, P1, P2} -> ast_lib:mk_intersection([ty_of_pat(Env, P1), ty_of_pat(Env, P2)]);
+        {match, _L, P1, P2} ->
+            ast_lib:mk_intersection([ty_of_pat(Env, P1, Mode), ty_of_pat(Env, P2, Mode)]);
         {nil, _L} -> {empty_list};
         {cons, _L, P1, P2} ->
-            %% FIXME: this is wrong. It should be the union of {list, ty_of_pat(Env, P1)}
-            %% and ty_of_pat(Env, P2), intersected with type list(any()).
-            %% Accepting and potential types of list patterns are also wrong.
-            ast_lib:mk_intersection([{list, ty_of_pat(Env, P1)}, ty_of_pat(Env, P2)]);
+            Symtab = symtab:empty(),  % FIXME: need to use a real symtab below?
```
Comment (albsch) [inline src/constr_gen.erl:378]: In the `typing:infer/2` function, there is a context `ctx` which includes the current Symtab. The call to `constr_gen:gen_constrs_fun_group/1` loses that symtab. The `ctx` type in `constr_gen` is missing the symtab. It should be added, and that symtab can (should) be used here.

File: src/constr_simp.erl [hunk 3]
```diff
@@ -150,6 +150,7 @@ simp_constr(Ctx, C) ->
                       Substs
                      );
                 report ->
+                    % FIXME: we miss an invocation of tally to get an substitution that we can apply to GammaI
```
Comment (albsch) [inline src/constr_simp.erl:153]: Is this tracked somewhere in an issue?

File: src/ety_main.erl [hunk 4]
```diff
@@ -31,6 +31,10 @@ parse_args(Args) ->
              "Add path to the front of Erlang's code path"},
          {load_end,   $z,   "pz",        string,
              "Add path to the end of Erlang's code path"},
+         {force,       $f,   "force",      undefined,
```
Comment (albsch) [inline src/ety_main.erl:34]: Is this the `--dirty` flag mentioned in #37 ?

**Review by albsch — CHANGES_REQUESTED** [review]
The correct symtab is needed for subtyping to work properly (in the future).

**Comment by albsch** [comment]
> @albsch This PR is ready for review.
> 
> One question: in `subtil.erl`, I find the following
> 
> ```erlang
> -spec is_subty(symtab:t(), ast:ty(), ast:ty()) -> boolean().
> is_subty(_Symtab, T1, T2) ->
> % TODO use symtab
> ```
> 
> What's the plan here? We need to symtab for expanding type definitions, right? Does it mean that the code currently does not work with user defined types?

Transforming the symtab in the `erlang_types` sub-library is a small implementation detail that will be added shortly in a PR.

File: src/constr_gen.erl (hunk 2 above)
Comment (skogsbaer) [reply src/constr_gen.erl:378]: fixed

File: src/ety_main.erl (hunk 4 above)
Comment (skogsbaer) [reply src/ety_main.erl:34]: yes

File: src/constr_gen.erl (hunk 1 above)
Comment (skogsbaer) [reply src/constr_gen.erl:11]: done

**Pull request merged**